### ISSUE TITLE
How to resume training from a saved checkpoint?

### ISSUE THREAD

**Issue by mzeidhassan** [modernmt/modernmt#445]
Hello MMT team,

I trained a model with 480k steps and I used the debug flag. I see in 'runtime/en-ro/tmp/training/neural_model' the following files:

model.ckpt-480000.data-00000-of-00002
model.ckpt-480000.data-00001-of-00002
model.ckpt-480000.index
model.ckpt-480000.meta

I would like to train this engine further for extra 90k steps. Is there a way to resume from the 480k ckpt file without the need to restart the whole training again from scratch? 

If this is possible, what would be the right command to do so?

I hope I will hear from you soon.

Thanks in advance for your support!
Mohamed

**Comment by EtienneAb3d** [comment]
If you didn't pack the model, simply use the exact command you launched to train, adding these options at the end:
`--resume --steps train_model`

Don't hesitate to make a backup of the whole folders of your model, both in the runtime and engines folders.

**Comment by mzeidhassan** [comment]
Thanks @EtienneAb3d for your help!

I tried this command:

`./mmt create en fr data/ --engine en-fr --train-steps 90000 -s train_model --resume`

But I got this error although I see the "checkpoint.json" file under '/tmp/training'. Maybe something wrong with my command. Any idea?

**Comment by EtienneAb3d** [comment]
Are you sure of your command ? In your first message, it was "en-ro", and in the latest "en-fr".
Otherwise.. certainly one of your previous commands deleted your tmp training data..

**Comment by mzeidhassan** [comment]
Yes, it's correct. I am testing it with en-fr, sorry about the en-ro mistake. 

It seems that something was deleted, and I believe also that since the initial training was completed successfully, you cannot resume it any longer.

I hope there is a way to iterate over an engine with more steps to tune it a bit if you wish. Or at least, to be able to use a saved checkpoint for further training later. Let's see what @davidecaroselli or @nicolabertoldi can say about this. Maybe, there is a way.

Thanks again @EtienneAb3d for your support. I really appreciate it.

**Comment by EtienneAb3d** [comment]
See #440:
1) to avoid the tmp training data to be deleted at the final "pack_model" step, you need to add the "--debug" option (in my case, I made a code modification before to know about this)
2) yo also need to manually remove all "pack_model" entries in the "checkpoint.json" file

**Comment by mzeidhassan** [comment]
Thanks @EtienneAb3d ! I will try to figure out what went wrong. I used the --debug flag though, but it seems the system I am doing the training on, once the training is over, some files are deleted.

**Comment by mzeidhassan** [comment]
Hi @davidecaroselli 
Is there a way to resume training from a particular saved checkpoint?

**Comment by nicolabertoldi** [comment]
@mzeidhassan 
If you used the flag `--debug``,  you should be able to resume the training., as already described in #440.
How does this procedure fail for you?

You mentioned that "some files are deleted". Could you please report which are these files?
or alternatively, could you please report:
- the content of directory `runtime/ENGINE/tmp/training` and its subdirectories
- the content of `runtime/ENGINE/tmp/training/checkpoint.json`

**Comment by davidecaroselli** [comment]
Hi @mzeidhassan 

just merged into master also the ability to start a training from an already saved checkpoint with `mmt create` option `--from-model`. Hope this is helpful!

**Comment by EtienneAb3d** [comment]
Not sure to understand @mzeidhassan, but, I think his question is a bit different. I also got this in mind by the past, but was never really blocked on it to ask: being several checkpoints 1..N, how to restart the training from one of these, C being between 1 and N ? Typically, it's interesting to pack C older than the last N, to see what would be the quality with an shorter training, ran when you was elsewhere and not ready to stop it at this time.

**Comment by mzeidhassan** [comment]
> * the content of directory `runtime/ENGINE/tmp/training` and its subdirectories
> * the content of `runtime/ENGINE/tmp/training/checkpoint.json`

@nicolabertoldi Here is a screen shot of the /tmp/training directory
![image](https://user-images.githubusercontent.com/7589948/53504956-2e5b6300-3a70-11e9-9e4b-cd65681b78fc.png)

and the content of checkpoint.json is this:

> ["clean_tms", "preprocess", "train_aligner", "write_config", "prepare_data", "train_model", "pack_model"]

**Comment by mzeidhassan** [comment]
@EtienneAb3d  Yes, I want to know if I am able to resume a training form a particular saved checkpoint. For example, at this path "/tmp/training/neural_model", I have:
![image](https://user-images.githubusercontent.com/7589948/53505272-c6594c80-3a70-11e9-8699-d7b48c18dc7c.png)

Is it possible to pack any of the saved checkpoints to base my engine on it? 
Can I resume training from any of these saved checkpoints?

**Comment by nicolabertoldi** [comment]
@mzeidhassan 
@EtienneAb3d 

I try to be clearer.
There are two cases:
A) you want to resume a training which was wrongly interrupted
B) you want to start a new training on (optionally) new data from a model, which have already been trained successfully

For case A, you have to follow instructions:
- remove all "pack_model" entries in the file `runtime/ENGINE/tmp/training/checkpoint.json`
- re-run the command 'mmt create with the additional parameter `--resume`
This works only if the previous training was run with the parameter `--debug`

For case B, you should use the latest option "--from-model MODELDIR".
where `MODELDIR` is the directory containing the model created previously.
In this case you can even choose to train on different data.

Hope this clarify the point and answer your questions.

**Comment by EtienneAb3d** [comment]
@nicolabertoldi , there is a case 'C', for example: 
If I have a checkpoint every 50K steps, after 500K steps, I have 10 checkpoints.
How can I resume, typically to pack it, checkpoint 3 done at 150K steps ?

**Comment by mzeidhassan** [comment]
@nicolabertoldi Just a follow-up question on case B:

Does that mean if my Model A has been trained with 'n' lines of data, if I want to train a Model B with 'n+m' lines of data, the Model B starts training from where Model A left (only the diff data, here only 'm' lines of data)? or does it retrain both 'n+m' data?

What is the impact on vocabulary? Would it expand to include new vocab from the newly added data?

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 
before considering how to work in case 'C', it is important to clarify how the "pack" steps works.

The "pack" step 
- looks for all checkpoints available in the training directory
- builds an average of them
- completes the training
- creates a working engine.
Moreover, it is important to stress that only the last K (=20) checkpoints are kept, while the oldest ones are removed.

So, one tricky way to "pack" the checkpoint 3 (and its predecessors, if any) could be:
- (re)move the next checkpoints (from 4 on)
- check and in case add in `runtime/ENGINE/tmp/training/checkpoint.json` the step 'train_model', to tell the system that to not continue the training
- check and in case remove in `runtime/ENGINE/tmp/training/checkpoint.json` the step 'pack_model', to tell the system that to not continue the training
- resume the training, which actually skips all steps but the "pack" step

**Comment by nicolabertoldi** [comment]
@mzeidhassan 
The case B is essentially a new training for which one can choose any training data. There is not relation with the data previously used.

**But**, with respect to the standard training, there are two main differences:
- the initial model is already trained, and not randomly initialized
- the vocabulary is **not** extended, but the one of the loaded model is exploited

**Comment by davidecaroselli** [comment]
Hi all!

Just a final note. With the new ModernMT v4, we have made this operation much more simple!
In fact you can resume a training at any time with `./mmt create [...] --resume`, **even if the previous training has completed successfully**.

The `--resume` flag in fact will force the training process to resume from the latest checkpoint, no matter how it stops before.
